src/source/custom_source.js

```diff
@@ -0,0 +1,319 @@
+
+    map: Map;
+    dispatcher: Dispatcher;
+    _loaded: boolean;
+
+    constructor(id: string, implementation: CustomSourceInterface<T>, dispatcher: Dispatcher, eventedParent: Evented) {
+        super();
+        this.id = id;
+        this.type = 'custom';
+        this.dataType = 'raster';
+        this.dispatcher = dispatcher;
+        this.setEventedParent(eventedParent);
+        this.implementation = implementation;
+
+        this.scheme = 'xyz';
+        this.minzoom = 0;
+        this.maxzoom = 22;
+        this.tileSize = 512;
+
+        this._loaded = false;
+        this.roundZoom = true;
+
+        if (!implementation) {
+            this.fire(new ErrorEvent(new Error(`Missing implementation for ${this.id} custom source`)));
+        }
+
+        if (!implementation.loadTile) {
+            this.fire(new ErrorEvent(new Error(`Missing loadTile implementation for ${this.id} custom source`)));
+        }
+
+        // $FlowFixMe[prop-missing]
+        implementation.update = this.update.bind(this);
+
+        // $FlowFixMe[prop-missing]
+        implementation.coveringTiles = this.coveringTiles.bind(this);
+
+        extend(this, pick(implementation, ['dataType', 'scheme', 'minzoom', 'maxzoom', 'tileSize', 'attribution', 'minTileCacheSize', 'maxTileCacheSize']));
+    }
+
+    serialize(): Source {
+        return pick(this, ['type', 'scheme', 'minzoom', 'maxzoom', 'tileSize', 'attribution']);
+    }
+
+    load() {
+        this._loaded = true;
+        this.fire(new Event('data', {dataType: 'source', sourceDataType: 'metadata'}));
+        this.fire(new Event('data', {dataType: 'source', sourceDataType: 'content'}));
+    }
+
+    loaded(): boolean {
+        return this._loaded;
+    }
+
+    onAdd(map: Map): void {
+        this.map = map;
+        this._loaded = false;
+        this.fire(new Event('dataloading', {dataType: 'source'}));
+
+        if (this.implementation.onAdd) {
+            this.implementation.onAdd(map, this.load);
+        } else {
+            this.load();
+        }
+    }
+
+    onRemove(map: Map): void {
+        if (this.implementation.onRemove) {
+            this.implementation.onRemove(map);
+        }
+    }
+
+    loadTile(tile: Tile, callback: Callback<void>) {
+        const {x, y, z} = tile.tileID.canonical;
+        const controller = new window.AbortController();
+        const signal = controller.signal;
+
+        // $FlowFixMe[prop-missing]
+        tile.request = this.implementation.loadTile({x, y, z}, {signal})
+            .then(tileLoaded.bind(this))
+            .catch(error => {
+                // silence AbortError and 404 errors
+                if (error.code === 20 || error.code === 404) return;
+                callback(error);
+            });
+
+        // $FlowFixMe[prop-missing]
+        tile.request.cancel = () => controller.abort();
+
+        function tileLoaded(data) {
+            delete tile.request;
+
+            if (tile.aborted) return callback(null);
```

Comment:
If the tile was aborted, should we still be calling this callback?

Reply:
This is what `raster_tile_source.js` does now https://github.com/mapbox/mapbox-gl-js/blob/main/src/source/raster_tile_source.js#L118-L120

Reply:
We need to notify the SourceCache that the source didn't return a tile in that case